Ensure Process.ProcessName doesn't change when setting Thread.Name on Linux

tmds opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/33673

This issue is a side-effect of adding support for setting thread names on Linux in https://github.com/dotnet/coreclr/pull/27182.

cc @janvorli @lpereira @SteveL-MSFT

tmds committed:
Add ThreadNameDoesNotAffectProcessName test

tmds committed:
Merge remote-tracking branch 'upstream/master' into process_name

tmds commented:
I've added only a test for now to let it fail in CI.
Next, I'll add a commit to ignore requests to set the main thread name.

stephentoub commented on src/libraries/System.Threading.Thread/tests/ThreadTests.cs:
```diff
@@ -647,6 +647,19 @@ public static void NameTest()
+            {
+                const string threadName = "my-thread";
+                Thread.CurrentThread.Name = threadName;
+                Assert.NotEqual(threadName, Process.GetCurrentProcess().ProcessName);
```
```suggestion
                const string ThreadName = "my-thread";
                Thread.CurrentThread.Name = ThreadName;
                Assert.NotEqual(ThreadName, Process.GetCurrentProcess().ProcessName);
                Assert.Equal(ThreadName, Thread.CurrentThread.Name);
```

stephentoub replied:
Do we have tests that change a non-main thread name and validate it roundtrips?

tmds replied:
By rountrip you mean: _when we ask the OS the thread name after setting it_?
I had a look in the `ThreadTests`, and there doesn't seem to be such a test.

The Thread.Name is considered a write-once value. It is stored in a field which the getter returns. There is no code to get the name from the OS.

stephentoub replied:
> By rountrip you mean: when we ask the OS the thread name after setting it?

Yup.  Mind adding one?

stephentoub replied:
I meant literally just set the property and then get the property.

tmds replied:
ok. There is a `NameTest` that does it for a non-main thread.
I'll add it here to cover the main-thread case also.

danmosemsft commented on src/libraries/System.Threading.Thread/tests/ThreadTests.cs:
```diff
@@ -647,6 +647,19 @@ public static void NameTest()
+            RemoteExecutor.Invoke(() =>
+            {
+                const string threadName = "my-thread";
+                Thread.CurrentThread.Name = threadName;
```
Nit, it's a bit pedantic, but this doesn't verify that the thread name didn't change to something else (eg empty). You might save it before the set , and verify it after. Instead of asserting not equal.

tmds replied:
`Thread.Name` is a write-once property that gets stored in a field. It never reads something from the OS.

danmosemsft replied:
I was essentially making the same suggestion as @stephentoub

tmds replied:
I realized when I was giving similar replies :smile:

tmds committed:
Assert Main Thread Name get returns set value.

tmds committed:
Ignore requests to set main thread name.

tmds commented on src/coreclr/src/pal/src/thread/thread.cpp:
```diff
@@ -1641,6 +1641,13 @@ CorUnix::InternalSetThreadDescription(
 
     pTargetThread->Lock(pThread);
 
+    // Ignore requests to set the main thread name because
```
I put this in pal, because it is a good place for calling `getpid`.

1. I'm not sure whether this needs to be under `->Lock`.
2. When looking at the code above `->Lock` I noticed `goto InternalSetThreadDescriptionExit;` is also used, though I don't understand how it knows when it shouldn't `->Unlock`.

danmosemsft replied:
> I don't understand how it knows when it shouldn't ->Unlock.

This may be a bug. `InternalLeaveCriticalSection` would no-op (I think) so it would not be noticed unless we followed this codepath with a checked runtime so the asserts fired?

tmds replied:
I looked into it a bit more. It is not a bug: `pTargetThread` will be `null`, and `->Unlock` isn't called.

am11 commented on src/coreclr/src/pal/src/thread/thread.cpp:
```diff
@@ -1641,6 +1641,13 @@ CorUnix::InternalSetThreadDescription(
 
+    // Ignore requests to set the main thread name because
+    // it causes the value returned by Process.ProcessName to change.
+    if (pTargetThread->GetThreadId() == getpid())
```
```suggestion
    if (pTargetThread->GetThreadId() == (unsigned)getpid())
```
will fix the GCC error.

tmds replied:
I've added a `pid_t` cast on `GetThreadId`.

tmds committed:
Fix GCC compilation

tmds commented:
This should be good to merge.
CI failures are unrelated.

mangod9 commented:
@tmds is this just waiting to be merged or is there anything else pending? Thx.

stephentoub commented:
Just waiting to be merged.

danmosemsft commented:
Simple cause: the test is failing on Mono
```
Expected: Not "my-thread"
Actual:   "my-thread"
   at System.Threading.Threads.Tests.ThreadTests.<>c.<ThreadNameDoesNotAffectProcessName>b__27_0() in /_/src/libraries/System.Threading.Thread/tests/ThreadTests.cs:line 664
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) in /_/src/mono/netcore/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.cs:line 359
```

Commit:
fix mono

danmosemsft commented:
I pushed a commit just to see whether it will fix that (I didn't test it yet .. feel free to discard)

Commit:
Revert "fix mono"

This reverts commit 376bbe57adbef5fbe265359f016b26c4928e17a2.

danmosemsft commented:
Nope - can't compare getpid() with a pthread_t. I'm not familar with pthreads.

lambdageek commented on src/mono/mono/utils/mono-threads-posix.c:
```diff
@@ -282,6 +282,13 @@ mono_native_thread_set_name (MonoNativeThreadId tid, const char *name)
+	 * it causes the value returned by Process.ProcessName to change.
+	 */
+	if (tid == getpid())
+		return;
```
```suggestion
#if defined(__linux__)
		if (mono_native_thread_os_id_get () == (guint64)getpid ())
			return;
#endif
```

I think this would work.

lambdageek replied:
If it doesn't work, please disable the test for Mono and open an issue for us

danmosemsft replied:
I don't have a Linux machine in a good state to verify it right now, but it does no harm to push and see...

tmds replied:
`getpid` and `pthread_t` (aka `MonoNativeThreadId`) can't be compared directly. On Linux, `pthread_t` points to some user-space struct about the thread. While `getpid` is from pid namespace that is maintained by the kernel. For the `pthread_t` there is a corresponding `pid_t`, but there is no public API to get it (from a `pthread_t`). Instead, we can call Linux specific `gettid` when the thread is started and store that value, so we can use it later for this comparison.

I can look at changing the mono implementation. I won't get to it this week.

lambdageek replied:
@tmds That's why I suggested `mono_native_thread_os_id_get ()` which just returns the tid on Linux.

Commit:
attempt 2

danmosemsft commented:
Thanks, @tmds. Another option is to disable the test for Mono for now and open an issue.

lambdageek commented on src/mono/mono/utils/mono-threads-posix.c:
```diff
@@ -282,6 +282,15 @@ mono_native_thread_set_name (MonoNativeThreadId tid, const char *name)
+	 * Ignore requests to set the main thread name because
+	 * it causes the value returned by Process.ProcessName to change.
+	 */
+	if (tid == (guint64)getpid ())
+		return;
```
```suggestion
	if (mono_native_thread_os_id_get () == (guint64)getpid ())
		return;
```

lambdageek committed:
[mono] (Linux) check native os thread id when setting thread name

Changing the name of the main thread changes the name of the entire process, which we don't want.

tmds commented on src/mono/mono/utils/mono-threads-posix.c:
```diff
@@ -282,6 +282,15 @@ mono_native_thread_set_name (MonoNativeThreadId tid, const char *name)
+	 * it causes the value returned by Process.ProcessName to change.
+	 */
+	if (mono_native_thread_os_id_get () == (guint64)getpid ())
+		return;
```
This is under `#ifdef __MACH__`.
And not aware of the thread we're trying to set the name for (`MonoNativeThreadId tid` argument).

lambdageek commented:
Bah. my commit is wrong for 2 reasons:
1. It's in the `__MACH__` section, which is the OSX version of the code, it should be in the `#elif defined (HAVE_PTHREAD_SETNAME_NP)` part which is where Linux should be.
2. It's checking the _current_ thread's OS id against `getpid()`.  Which is wrong. We need to map `MonoNativeThreadId` (`pthread_t`) to `pid_t`.  That won't be an easy fix.  Going to revert my change.

@tmds I'm going to revert my change, open a new GH Issue for Mono runtime, and then disable the test for Mono

lambdageek commented on src/mono/mono/utils/mono-threads-posix.c:
```diff
@@ -282,6 +282,15 @@ mono_native_thread_set_name (MonoNativeThreadId tid, const char *name)
 	if (tid != mono_native_thread_id_get ())
 		return;
 
+#if defined(__linux__)
+	/*
+	 * Ignore requests to set the main thread name because
+	 * it causes the value returned by Process.ProcessName to change.
+	 */
+	if (mono_native_thread_os_id_get () == (guint64)getpid ())
+		return;
+#endif
+
```
```suggestion
```

lambdageek commented on src/libraries/System.Threading.Thread/tests/ThreadTests.cs:
```diff
@@ -647,6 +647,20 @@ public static void NameTest()
             });
         }
 
+        [Fact]
+        public static void ThreadNameDoesNotAffectProcessName()
```
```suggestion
        [Fact]
        [ActiveIssue ("https://github.com/dotnet/runtime/issues/35908", TestRuntimes.Mono)]
        public static void ThreadNameDoesNotAffectProcessName()
```

lambdageek reviewed (commented):
Revert the Mono changes and disable the test until #35908 is fixed

lambdageek committed:
Disable ThreadNameDoesNotAffectProcessName on Mono

until https://github.com/dotnet/runtime/issues/35908 is fixed

lambdageek committed:
Revert mono change

lambdageek commented:
@tmds Sorry about the mess.

akoeplinger commented:
Test failure in `runtime (CoreCLR Pri0 Runtime Tests Run Windows_NT x86 checked)` is unrelated: https://github.com/dotnet/runtime/issues/35877
